fw/ja5_conf.c

```diff
@@ -0,0 +1,243 @@
+	rcu_read_unlock_bh();
+
+	return entry;
+}
+
+static void
+tls_put_ja5_hash_entry(TlsJa5HashEntry *entry)
+{
+	if (entry && !atomic64_dec_return(&entry->refcnt))
+		kfree(entry);
+}
+
+u64
+tls_get_ja5_conns_limit(TlsJa5t fingerprint)
+{
+	u64 res = -1;
+	TlsJa5HashEntry *e = tls_get_ja5_hash_entry(fingerprint);
+
+	if (e) {
+		res = e->conns_per_sec;
+		tls_put_ja5_hash_entry(e);
+	}
+
+	return res;
+}
+
+u64
+tls_get_ja5_recs_limit(TlsJa5t fingerprint)
+{
+	u64 res = -1;
+	TlsJa5HashEntry *e = tls_get_ja5_hash_entry(fingerprint);
+
+	if (e) {
+		res = e->records_per_sec;
+		tls_put_ja5_hash_entry(e);
+	}
+
+	return res;
+}
+
+size_t
+tls_get_ja5_storage_size(void)
+{
+	size_t res = 0;
+
+	if (tls_filter_cfg) {
+		rcu_read_lock_bh();
+		res = rcu_dereference_bh(tls_filter_cfg)->storage_size;
+		rcu_read_unlock_bh();
+	}
+
+	return res;
+}
+
+int
+ja5_cfgop_handle_hash_entry(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	TlsJa5t hash;
+	u32 conns_per_sec;
+	u32 recs_per_sec;
+	TlsJa5HashEntry *he;
+	u64 key;
+
+	BUILD_BUG_ON(sizeof(hash) > sizeof(u64));
+	TFW_CFG_CHECK_VAL_EQ_N(3, cs, ce);
+	TFW_CFG_CHECK_NO_ATTRS(cs, ce);
+
+	if (kstrtouint(ce->vals[1], 10, &conns_per_sec)) {
+		T_ERR_NL("Failed to parse hash entry in ja5 section: "
+			"invalid connections per second value %s", ce->vals[1]);
+		return -EINVAL;
+	}
+
+	if (kstrtouint(ce->vals[2], 10, &recs_per_sec)) {
+		T_ERR_NL("Failed to parse hash entry in ja5 section: "
+			"invalid records per second value %s", ce->vals[2]);
+		return -EINVAL;
+	}
+
+	if (kstrtou64(ce->vals[0], 16, (u64 *)&hash)) {
+		T_ERR_NL("Failed to parse hash entry in ja5 section: "
+			"invalid hash value %s", ce->vals[0]);
+		return -EINVAL;
+	}
+
+	if (!(he = kmalloc(sizeof(TlsJa5HashEntry), GFP_KERNEL)))
+		return -ENOMEM;
+
+	he->ja5_hash = hash;
+	he->conns_per_sec = conns_per_sec;
+	he->records_per_sec = recs_per_sec;
+	INIT_HLIST_NODE(&he->hlist);
+	he->refcnt.counter = 1;
+
+	key = hash_calc((char *)&hash, sizeof(hash));
+	hash_add(tls_filter_cfg_reconfig->hashes, &he->hlist, key);
+
+	return 0;
+}
+
+int
+ja5_cfgop_begin(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	BUG_ON(tls_filter_cfg_reconfig);
+	TFW_CFG_CHECK_VAL_EQ_N(0, cs, ce);
+	TFW_CFG_CHECK_ATTR_LE_N(1, cs, ce);
+
+	if (!(tls_filter_cfg_reconfig =
+		kzalloc(sizeof(TlsJa5FilterCfg), GFP_KERNEL)))
+		return -ENOMEM;
+
+	if (ce->attr_n == 1) {
+		if (strcasecmp(ce->attrs[0].key, "storage_size")) {
+			T_ERR_NL("Failed to parse ja5 section: "
+			"invalid attribute %s", ce->attrs[0].key);
+			return -EINVAL;
+		}
+
+		if (kstrtoul(ce->attrs[0].val, 10,
+			&tls_filter_cfg_reconfig->storage_size)) {
+			T_ERR_NL("Failed to parse ja5 section: "
+				"invalid storage_size value");
+			return -EINVAL;
+		}
+	} else {
+		tls_filter_cfg_reconfig->storage_size =
+			TLS_JA5_DEFAULT_STORAGE_SIZE;
+	}
+
+	return 0;
+}
+
+static void
+free_cfg(TlsJa5FilterCfg *cfg)
+{
+	u32 bkt_i;
+	TlsJa5HashEntry *entry;
+
+	BUG_ON(!cfg);
+
+	hash_for_each(cfg->hashes, bkt_i, entry, hlist)
```

Comment:
Should we use [hash_for_each_safe](https://elixir.bootlin.com/linux/v6.12.4/C/ident/hash_for_each_safe) ?
As read in documentation hash_for_each_safe - iterate over a hashtable safe against removal of hash entry

Reply:
done